File: io-hotmoka-service/src/main/java/io/hotmoka/service/models/responses/JarStoreTransactionResponseModel.java
package io.hotmoka.service.models.responses;

import java.util.List;
import java.util.stream.Collectors;

import io.hotmoka.beans.responses.JarStoreNonInitialTransactionResponse;
import io.hotmoka.service.models.updates.UpdateModel;

public abstract class JarStoreTransactionResponseModel extends TransactionResponseModel {

    /**
     * The updates resulting from the execution of the transaction.
     */
    public List<UpdateModel> updates;

    /**
     * The amount of gas consumed by the transaction for CPU execution.
     */
    public String gasConsumedForCPU;

    /**
     * The amount of gas consumed by the transaction for RAM allocation.
     */
    public String gasConsumedForRAM;

    /**
     * The amount of gas consumed by the transaction for storage consumption.
     */
    public String gasConsumedForStorage;

    protected JarStoreTransactionResponseModel(JarStoreNonInitialTransactionResponse response) {
       this.updates = response.getUpdates().map(UpdateModel::new).collect(Collectors.toList());
       this.gasConsumedForCPU = response.gasConsumedForCPU.toString();
       this.gasConsumedForRAM = response.gasConsumedForRAM.toString();
       this.gasConsumedForStorage = response.gasConsumedForStorage.toString();
    }

    protected JarStoreTransactionResponseModel() {}
}

File: io-hotmoka-network-thin-client/io-hotmoka-network-thin-client-kt/src/main/kotlin/io/hotmoka/network/thin/client/models/responses/JarStoreTransactionResponseModel.kt
package io.hotmoka.network.thin.client.models.responses

import io.hotmoka.network.thin.client.exceptions.TransactionException
import io.hotmoka.network.thin.client.models.updates.UpdateModel
import io.hotmoka.network.thin.client.models.values.TransactionReferenceModel
import kotlin.jvm.Throws

abstract class JarStoreTransactionResponseModel(
        /**
         * The updates resulting from the execution of the transaction.
         */
        val update: List<UpdateModel>,
        /**
         * The amount of gas consumed by the transaction for CPU execution.
         */
        val gasConsumedForCPU: String,
        /**
         * The amount of gas consumed by the transaction for RAM allocation.
         */
        val gasConsumedForRAM: String,
        /**
         * The amount of gas consumed by the transaction for storage consumption.
         */
        val gasConsumedForStorage: String
) : TransactionResponseModel() {


        /**
         * Yields the outcome of the execution having this response, performed
         * at the given transaction reference.
         *
         * @param transactionReference the transaction reference
         * @return the outcome
         * @throws TransactionException if the outcome of the transaction is this exception
         */
        @Throws(TransactionException::class)
        abstract fun getOutcomeAt(transactionReference: TransactionReferenceModel): TransactionReferenceModel
}